src/mscorlib/src/System/Runtime/ExceptionServices/ExceptionServicesCommon.cs

```diff
@@ -125,9 +125,14 @@ public Exception SourceException
         // the original thread.
         public void Throw()
         {
-            // Restore the exception dispatch details before throwing the exception.
+            throw GetThrowableException();
+        }
+
+        public Exception GetThrowableException()
+        {
+            // Restore the exception dispatch details before allowing this exception to be thrown.
             m_Exception.RestoreExceptionDispatchInfo(this);
-            throw m_Exception;
+            return m_Exception;
```

Comment:
Someone more familiar with the innards of exceptions would need to pipe in on whether handing out an exception object like this with the EDI info restored is acceptable.

Reply:
It might be the issue changing the line numbers; although they didn't seem completely right in the first place (maybe release mode?)

e.g. 
AsyncMethod always at 27 in pre
AsyncMethod at 27, 22, 22, 22... in post

However the source is 23, 27, 27, 27, 27...